File: frontend/src/stores/notification.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import axios from 'axios';
import type { Notification, NotificationConfig } from '../types/notification'
import { API_CONFIG } from '../config/api';

export const useNotificationStore = defineStore('notification', () => {
  const notifications = ref<Notification[]>([])
  const total = ref(0)
  const loading = ref(false)
  const config = ref<NotificationConfig | null>(null)
  const _unreadCount = ref(0)

  // 获取通知配置
  const fetchConfig = async () => {
    try {
      const { data } = await axios.get(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.CONFIG}`)
      config.value = data
    } catch (error) {
      console.error('获取通知配置失败:', error)
    }
  }

  // 更新通知配置
  const updateConfig = async (newConfig: Partial<NotificationConfig>) => {
    try {
      await axios.put(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.CONFIG}`, newConfig)
      await fetchConfig()
      return true
    } catch (error) {
      console.error('更新配置失败:', error)
      return false
    }
  }

  // 测试通知配置
  const testConfig = async (type: 'email' | 'wechat', config: Record<string, any>) => {
    try {
      await axios.post(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.TEST}`, { type, config })
      return true
    } catch (error) {
      console.error('测试发送失败:', error)
      return false
    }
  }

  // 获取通知历史
  const fetchHistory = async (params: {
    type?: 'scan' | 'ip' | 'policy'
    status?: 'read' | 'unread'
    dateRange?: [Date, Date]
    page: number
    per_page: number
  }) => {
    try {
      loading.value = true
      const { data } = await axios.get(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.HISTORY}`, { params })
      notifications.value = data.notifications
      total.value = data.total
    } catch (error) {
      console.error('获取通知历史失败:', error)
    } finally {
      loading.value = false
    }
  }

  // 标记通知为已读
  const markAsRead = async (id: string) => {
    try {
      await axios.put(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.MARK_READ(id)}`)
      const notification = notifications.value.find(n => n.id === id)
      if (notification) {
        notification.read = true
      }
      await fetchUnreadCount()
      return true
    } catch (error) {
      console.error('标记已读失败:', error)
      return false
    }
  }

  // 标记所有通知为已读
  const markAllAsRead = async () => {
    try {
      await axios.put(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.MARK_ALL_READ}`)
      notifications.value.forEach(n => n.read = true)
      await fetchUnreadCount()
      return true
    } catch (error) {
      console.error('全部标记已读失败:', error)
      return false
    }
  }

  // 删除通知
  const deleteNotification = async (id: string) => {
    try {
      await axios.delete(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.DELETE(id)}`)
      notifications.value = notifications.value.filter(n => n.id !== id)
      total.value = total.value - 1
      await fetchUnreadCount()
      return true
    } catch (error) {
      console.error('删除通知失败:', error)
      return false
    }
  }

  // 清空所有通知
  const clearAll = async () => {
    try {
      await axios.delete(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.CLEAR_ALL}`)
      notifications.value = []
      total.value = 0
      _unreadCount.value = 0
      return true
    } catch (error) {
      console.error('清空通知失败:', error)
      return false
    }
  }

  // 获取未读通知数量
  const fetchUnreadCount = async () => {
    try {
      const { data } = await axios.get(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.NOTIFICATION.UNREAD_COUNT}`)
      _unreadCount.value = data.count
    } catch (error) {
      console.error('获取未读通知数量失败:', error)
      _unreadCount.value = 0
    }
  }

  // 未读通知数量（对外暴露）
  const unreadCount = computed(() => _unreadCount.value)

  return {
    notifications,
    total,
    loading,
    config,
    unreadCount,
    fetchConfig,
    updateConfig,
    testConfig,
    fetchHistory,
    markAsRead,
    markAllAsRead,
    deleteNotification,
    clearAll,
    fetchUnreadCount
  }
}) 

File: backend/app/api/v1/notification/notification.py
from flask import Blueprint, jsonify, request, current_app
from app.core.security.auth import token_required
from app.models.models import db, Notification
from app.core.utils.helpers import log_action_to_db
import smtplib # Import smtplib

notification_bp = Blueprint('notification', __name__)

@notification_bp.route('/notification/config', methods=['GET'])
@token_required
def get_config(current_user):
    """获取通知配置"""
    try:
        configs = current_app.notification_manager.get_all_configs()
        return jsonify(configs)
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/config', methods=['PUT'])
@token_required
def update_config(current_user):
    """更新通知配置"""
    try:
        config = request.json
        for key, value in config.items():
            current_app.notification_manager.update_config(current_user.id , key, value)
        return jsonify({'message': '配置更新成功'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/test', methods=['POST'])
@token_required
def test_config(current_user):
    """测试通知配置"""
    try:
        data = request.json
        type = data.get('type')
        config = data.get('config')

        if type == 'email':
            smtp_config = {
                'server': config.get('smtpServer'),
                'username': config.get('smtpUsername'),
                'password': config.get('smtpPassword'),
                'from': config.get('smtpFrom')
            }
            try:
                current_app.notification_manager.email_notifier.send(
                    current_user.email,
                    '这是一条测试通知',
                    smtp_config
                )
                return jsonify({'message': '测试发送成功'}) # Return success here
            except smtplib.SMTPAuthenticationError as e:
                current_app.logger.error(f"Failed to send email notification (Authentication Error): {str(e)}")
                return jsonify({'error': 'SMTP认证失败，请检查用户名和密码。'}), 400
            except smtplib.SMTPException as e:
                current_app.logger.error(f"Failed to send email notification (SMTP Error): {str(e)}")
                return jsonify({'error': '邮件服务器错误：' + str(e)}), 500
            except Exception as e:
                current_app.logger.error(f"Failed to send email notification (General Error): {str(e)}")
                return jsonify({'error': '邮件发送失败：' + str(e)}), 500

        elif type == 'wechat':
            try:
                current_app.notification_manager.wechat_notifier.send(
                    current_user.wechat_id,
                    '这是一条测试通知',
                    config.get('webhookUrl')
                )
                return jsonify({'message': '测试发送成功'}) # Moved this
            except Exception as e:
                current_app.logger.error(f"Failed to send wechat notification: {str(e)}")
                return jsonify({'error': '微信通知发送失败：' + str(e)}), 500

        return jsonify({'error': '无效的通知类型'}), 400 # Handle invalid type
    except Exception as e: # Catch any other unexpected errors
        current_app.logger.error(f"An unexpected error occurred during notification test: {str(e)}")
        return jsonify({'error': '服务器内部错误：' + str(e)}), 500

@notification_bp.route('/notification/history', methods=['GET'])
@token_required
def get_history(current_user):
    """获取通知历史"""
    try:
        page = request.args.get('page', 1, type=int)
        per_page = request.args.get('per_page', 10, type=int)
        type = request.args.get('type')
        status = request.args.get('status')
        
        query = Notification.query.filter_by(user_id=current_user.id)
        
        if type:
            query = query.filter_by(type=type)
        if status:
            query = query.filter_by(read=(status == 'read'))
            
        pagination = query.order_by(Notification.created_at.desc()).paginate(
            page=page, per_page=per_page
        )
        
        return jsonify({
            'notifications': [notification.to_dict() for notification in pagination.items],
            'total': pagination.total
        })
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/<string:id>/read', methods=['PUT'])
@token_required
def mark_as_read(current_user, id):
    """标记通知为已读"""
    try:
        notification = Notification.query.filter_by(id=id, user_id=current_user.id).first()
        if not notification:
            return jsonify({'error': '通知不存在'}), 404
            
        notification.read = True
        db.session.commit()
        return jsonify({'message': '标记已读成功'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/read-all', methods=['PUT'])
@token_required
def mark_all_as_read(current_user):
    """标记所有通知为已读"""
    try:
        Notification.query.filter_by(user_id=current_user.id, read=False).update({'read': True})
        db.session.commit()
        return jsonify({'message': '全部标记已读成功'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/<string:id>', methods=['DELETE'])
@token_required
def delete_notification(current_user, id):
    """删除通知"""
    try:
        notification = Notification.query.filter_by(id=id, user_id=current_user.id).first()
        if not notification:
            return jsonify({'error': '通知不存在'}), 404
            
        notification.deleted = True
        db.session.commit()
        return jsonify({'message': '删除成功'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/clear-all', methods=['DELETE'])
@token_required
def clear_all(current_user):
    """清空所有通知"""
    try:
        notification = Notification.query.filter_by(user_id=current_user.id).all()
        notification.deleted = True
        db.session.commit()
        return jsonify({'message': '清空成功'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@notification_bp.route('/notification/unread-count', methods=['GET'])
@token_required
def get_unread_count(current_user):
    """获取当前用户的未读通知数量"""
    try:
        unread_count = Notification.query.filter_by(user_id=current_user.id, read=False).count()
        return jsonify({'count': unread_count})
    except Exception as e:
        current_app.logger.error(f"获取未读通知数量失败: {str(e)}")
        return jsonify({'error': '获取未读通知数量失败'}), 500 